app/src/main/java/com/alexbar10/cryptotrack/ui/cryptoDetails/CryptocurrencyDetailsViewModel.kt

```diff
@@ -0,0 +1,175 @@
+package com.alexbar10.cryptotrack.ui.cryptoDetails
+
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.alexbar10.cryptotrack.database.entities.OrderEntity
+import com.alexbar10.cryptotrack.database.repo.CryptoDBRepo
+import com.alexbar10.cryptotrack.domain.Cryptocurrency
+import com.alexbar10.cryptotrack.domain.Failure
+import com.alexbar10.cryptotrack.domain.Order
+import com.alexbar10.cryptotrack.domain.OrderDetailItem
+import com.alexbar10.cryptotrack.domain.OrderResponse
+import com.alexbar10.cryptotrack.domain.Success
+import com.alexbar10.cryptotrack.networking.repo.CryptocurrenciesRepo
+import com.alexbar10.cryptotrack.networking.repo.CryptosRxRepo
+import dagger.hilt.android.lifecycle.HiltViewModel
+import io.reactivex.Observer
+import io.reactivex.android.schedulers.AndroidSchedulers
+import io.reactivex.disposables.Disposable
+import io.reactivex.schedulers.Schedulers
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@HiltViewModel
+class CryptocurrencyDetailsViewModel @Inject constructor(
+    private val cryptocurrenciesRepo: CryptocurrenciesRepo,
+    private val cryptosRxRepo: CryptosRxRepo,
+    private val repoDBRepo: CryptoDBRepo
+) : ViewModel() {
+
+    private val orderTypeAsk = 1
+    private val orderTypeBid = 2
+    private val _loadingLiveData = MutableLiveData<Boolean>()
+    val loadingLiveData: LiveData<Boolean> get() = _loadingLiveData
+
+    private val _errorLiveData = MutableLiveData<Throwable?>()
+    val errorLiveData: LiveData<Throwable?> get() = _errorLiveData
+
+    private val _cryptocurrencyOrderLiveData = MutableLiveData<OrderResponse>()
+    val cryptocurrencyOrderLiveData: LiveData<OrderResponse> get() = _cryptocurrencyOrderLiveData
+
+    // Using RxJava
+    fun getOrderRxFor(cryptocurrency: Cryptocurrency) {
+        _loadingLiveData.postValue(true)
+
+        cryptosRxRepo.getOrderFor(cryptocurrency.book)
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribeOn(Schedulers.io())
+            .subscribe(handleResponse(cryptocurrency))
```

Comment:
We need to be cautious with Rx with Android, the observers can potentially be kept in-memory even when no longer needed due to their references and because they're not tied to the ViewModel or any Android lifecycle component. For guaranteeing we clean-up all resources we can use a `disaposable`. As you see, we just wrap the subscription inside the disaposable `add` and we clean the resource in the `onCleared` method. In case you add more observers to the same Disposable. That will guarantee all Observers are disposed/cleared whenever the ViewModel is destroyed 👍 

```
val disposable = CompositeDisposable()

// Using RxJava
fun getOrderRxFor(cryptocurrency: Cryptocurrency) {
    _loadingLiveData.postValue(true)
    
    disposable.add(
        cryptosRxRepo.getOrderFor(cryptocurrency.book)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(handleResponse(cryptocurrency))
    )
}

override fun onCleared() {
    super.onCleared()
    disposable.clear()
  }
```